include/util/packed_vector.hpp

```diff
@@ -31,188 +35,353 @@ inline void write(storage::io::FileWriter &writer,
+inline T get_lower_half_value(WordT word,
+                              WordT mask,
+                              std::uint8_t offset,
+                              typename std::enable_if_t<std::is_integral<T>::value> * = 0)
+{
+    return static_cast<T>((word & mask) >> offset);
+}
+
+template <typename WordT, typename T>
+inline T
+get_lower_half_value(WordT word, WordT mask, std::uint8_t offset, typename T::value_type * = 0)
+{
+    return T{static_cast<typename T::value_type>((word & mask) >> offset)};
+}
+
+template <typename WordT, typename T>
+inline T get_upper_half_value(WordT word,
+                              WordT mask,
+                              std::uint8_t offset,
+                              typename std::enable_if_t<std::is_integral<T>::value> * = 0)
+{
+    return static_cast<T>((word & mask) << offset);
+}
+
+template <typename WordT, typename T>
+inline T
+get_upper_half_value(WordT word, WordT mask, std::uint8_t offset, typename T::value_type * = 0)
+{
+    return T{static_cast<typename T::value_type>((word & mask) << offset)};
+}
+
+template <typename WordT, typename T>
+inline WordT set_lower_value(WordT word, WordT mask, std::uint8_t offset, T value)
+{
+    return (word & ~mask) | (static_cast<WordT>(value) << offset);
+}
+
+template <typename WordT, typename T>
+inline WordT set_upper_value(WordT word, WordT mask, std::uint8_t offset, T value)
+{
+    return (word & ~mask) | (static_cast<WordT>(value) >> offset);
+}
+
 template <typename T, std::size_t Bits, storage::Ownership Ownership> class PackedVector
 {
+    using WordT = std::uint64_t;
+
     // This fails for all strong typedef types
     // static_assert(std::is_integral<T>::value, "T must be an integral type.");
-    static_assert(sizeof(T) <= sizeof(std::uint64_t), "Maximum size of type T is 8 bytes");
+    static_assert(sizeof(T) <= sizeof(WordT), "Maximum size of type T is 8 bytes");
     static_assert(Bits > 0, "Minimum number of bits is 0.");
-    static_assert(Bits <= sizeof(std::uint64_t) * CHAR_BIT, "Maximum number of bits is 64.");
-
-    static const constexpr std::size_t ELEMSIZE = sizeof(std::uint64_t) * CHAR_BIT;
-    static const constexpr std::size_t PACKSIZE = Bits * ELEMSIZE;
+    static_assert(Bits <= sizeof(WordT) * CHAR_BIT, "Maximum number of bits is 64.");
 
+    static constexpr std::size_t WORD_BITS = sizeof(WordT) * CHAR_BIT;
+    // number of elements per block, use the number of bits so we make sure
+    // we can devide the total number of bits by the element bis
   public:
-    using value_type = T;
-
-    /**
-     * Returns the size of the packed vector datastructure with `elements` packed elements (the size
-     * of
-     * its underlying uint64 vector)
-     */
-    inline static std::size_t elements_to_blocks(std::size_t elements)
-    {
-        return std::ceil(static_cast<double>(elements) * Bits / ELEMSIZE);
-    }
+    static constexpr std::size_t BLOCK_ELEMENTS = WORD_BITS;
```

Comment:
This is to make the total number of bits in a block devisable by the number of bits in a word. For example if every record has 22bits then 22*64bits => 22 words per block. Technically we could also try to compute the smallest common multiple of the number of bits in a record and the number of bits in a word, but this was simpler.